pkg/api/defaults.go

```diff
@@ -508,6 +512,83 @@ func (p *Properties) setHostedMasterProfileDefaults() {
 	p.HostedMasterProfile.Subnet = DefaultKubernetesMasterSubnet
 }
 
+func (p *Properties) setCustomCloudProfileDefaults() {
+
+	if p.IsAzureStackCloud() {
+		//Set default value for ResourceManagerVMDNSSuffix
+		AzureStackCloudSpec.EndpointConfig.ResourceManagerVMDNSSuffix = p.CustomCloudProfile.Environment.ResourceManagerVMDNSSuffix
```

Comment:
good idea, I will update the code.

Reply:
default value of AzureEnvironmentSpecConfig is not need under vlabs, I removed them.

Reply:
I don't see the change applied, it still looks like this method is mutating the "global variables" `AzureStackCloudSpec` and `AzureCloudSpecEnvMap`.

@tariq1890 do you have any thoughts/suggestions on a preferable pattern for what we're doing here?

Reply:
AzureCloudSpecEnvMap is using by the following function.
After setting the default value, the following function will be invoked to populate parameter values and certificate FQDN. 

So until we have other pattern to retrieve the value, I would need overwrite the global variables firstly and it will be used for the following functions. 

@jackfrancis @tariq1890 , Please let me know if you have ideas for different pattern, I am happy to make the changes and veirfy them.

func (cs *ContainerService) GetCloudSpecConfig() AzureEnvironmentSpecConfig {
	targetEnv := helpers.GetCloudTargetEnv(cs.Location, cs.Properties.GetCustomCloudName())
	return AzureCloudSpecEnvMap[targetEnv]
}

func FormatAzureProdFQDNByLocation(fqdnPrefix string, location string, cloudName string) string {
	targetEnv := helpers.GetCloudTargetEnv(location, cloudName)
	FQDNFormat := AzureCloudSpecEnvMap[targetEnv].EndpointConfig.ResourceManagerVMDNSSuffix
	return fmt.Sprintf("%s.%s."+FQDNFormat, fqdnPrefix, location)
}

Reply:
Sorry, my bad, I missed to address your comments.
Code is updated right now.